Python/fileutils.c

```diff
@@ -2450,6 +2470,47 @@ _Py_set_blocking(int fd, int blocking)
     return -1;
 }
 #else   /* MS_WINDOWS */
+int
+_Py_get_blocking(int fd)
+{
+    HANDLE handle;
+    DWORD mode;
+
+    handle = _Py_get_osfhandle(fd);
+    if (handle == INVALID_HANDLE_VALUE) {
+        return -1;
+    }
+
+    if (!GetNamedPipeHandleStateW(handle, &mode, NULL, NULL, NULL, NULL, 0)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
+    return mode & PIPE_WAIT;
+}
+
+int
+_Py_set_blocking(int fd, int blocking)
+{
+    HANDLE handle;
+    DWORD mode;
+
+    handle = _Py_get_osfhandle(fd);
+    if (handle == INVALID_HANDLE_VALUE) {
+        return -1;
+    }
+
+    if (!GetNamedPipeHandleStateW(handle, &mode, NULL, NULL, NULL, NULL, 0)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
+    mode |= blocking ? PIPE_WAIT : PIPE_NOWAIT;
+    if (!SetNamedPipeHandleState(handle, &mode, NULL, NULL)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
```

Comment:
Same for the GIL here, and my previous suggestion messed up the logic here as well. The `PIPE_NOWAIT` flag has to be cleared when `blocking` is true.

```suggestion
    DWORD mode;
    BOOL success;

    handle = _Py_get_osfhandle(fd);
    if (handle == INVALID_HANDLE_VALUE) {
        return -1;
    }

    Py_BEGIN_ALLOW_THREADS
    success = GetNamedPipeHandleStateW(handle, &mode,
                                       NULL, NULL, NULL, NULL, 0);
    if (success) {
        if (blocking) {
            mode &= ~PIPE_NOWAIT;
        }
        else {
            mode |= PIPE_NOWAIT;
        }
        success = SetNamedPipeHandleState(handle, &mode, NULL, NULL);
    }
    Py_END_ALLOW_THREADS

    if (!success) {
        PyErr_SetFromWindowsErr(0);
        return -1;
    }
```